src/util/system.cpp

```diff
@@ -974,17 +974,19 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
     SetEndOfFile(hFile);
 #elif defined(MAC_OSX)
     // OSX specific version
+    // NOTE: Contrary to other OS versions, the OSX version assumes that
+    // NOTE: offset is the size of the file.
```

Comment:
Wouldn't it make sense to add an assertion here? Implementing a function differently than documented on a OS could turn into a troubleshooting nightmare at some point.

Reply:
You would have to check file size, in which case you might as well do what @arvidn suggested. It depends on the impact of checking the file size (once per MB of data for rev files). I initially wrote code to do just that, but it felt wasteful. If people think it's fine, I can switch to this (or if there's a more optimized variant).

```diff
diff --git a/src/util/system.cpp b/src/util/system.cpp
index d9fc38c91..3edcf5bc3 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -974,11 +974,17 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
     SetEndOfFile(hFile);
 #elif defined(MAC_OSX)
     // OSX specific version
+    long pos = ftell(file);
+    fseek(file, 0L, SEEK_END);
+    long filesize = ftell(file);
+    fseek(file, pos, SEEK_SET);
+    if (offset + length <= filesize) return;
     fstore_t fst;
     fst.fst_flags = F_ALLOCATECONTIG;
     fst.fst_posmode = F_PEOFPOSMODE;
     fst.fst_offset = 0;
-    fst.fst_length = length; // mac os fst_length takes the # of free bytes to allocate, not desired file size
+    // mac os fst_length takes the # of free bytes to allocate, not desired file size
+    fst.fst_length = static_cast<off_t>(offset) + length - filesize;
     fst.fst_bytesalloc = 0;
     if (fcntl(fileno(file), F_PREALLOCATE, &fst) == -1) {
         fst.fst_flags = F_ALLOCATEALL;
```

Reply:
Yes, that seems wasteful.

Another solution would be to remove the `offset` parameter and define the function to always work relative to the end of the file, for all platforms. This also avoids using the function in the wrong away.

Reply:
I think the main reason why it is there in the first place is to avoid the seek part when possible. It basically functions as a cache approximating the file size. It will overshoot by a small amount (depending on the size of whatever is being written that doesn't currently fit), but it's mostly accurate, and I think for performance reasons this is done that way. After all, only the *preallocate* part overshoots - the `ftruncate` call adjusts the file size to the expected value (..except it doesn't -- see #17892).

We could of course remove the offset completely, but I think it would be an unnecessary performance loss with very little gain.

In practice, we end up over-pre-allocating up to the maximum possible value of an entry written into the respective files; for rev files, I assume this is usually in the few kbs range, and for blk files, I assume it's a couple of MBs since blocks are around that size. Analysis on this as below.

===

Given:
* `required` = bytes needed
* `position` = current position inside file
* `chunk_size` = the size of each chunk in the file (1 MB for rev files)
* `allocation` = current file allocation (pre-allocated space)
* `filesize` = current file *size*, i.e. whatever we truncated it to (this is identical to `allocation` on systems except OSX, and is in part mitigated by the fix here)
* `old_chunks = (position + chunk_size - 1) / chunk_size`
* `new_chunks = (position + required + chunk_size - 1) / chunk_size`

For the case where `new_chunks > old_chunks`, there is a discrepancy in `position` vs `filesize` that is between `0..chunk_size-1` (not greater, because the code never allocates more chunks than needed).
* The `offset` is set to the former, and thus the assumed file size is in the range `[filesize - chunk_offset + 1 .. filesize]`.
* The `length` is set to the value `new_chunks * chunk_size - position`, i.e. `position + required + chunk_size - 1 - position = required + chunk_size - 1`.

Results:
* Expectation: pre-allocate `required + chunk_size - 1` free bytes to get a file with size `position + required + chunk_size - 1` bytes, then truncate to that size.
* Result: pre-allocate `required + chunk_size - 1` free bytes to get a file with size `filesize + required + chunk_size - 1` bytes, then truncate to `position + required + chunk_size - 1` bytes (this is currently not functional on APFS).
* I.e. we end up with up to `chunk_size` more bytes in the pre-allocation stage.
* For the rev case, we end up over-pre-allocating up to 1 extra MB of space that we keep bumping (fairly randomly in the range 0..1024*1024, though in practice this value is capped at the size of the biggest entry we write into rev files).
* For the blk case, we over-pre-allocate up to 16 MB, but since blocks are capped at a few MBs, this ends up being a few MBs in practice.

If you spot any errors, let me know and I'll update.

Reply:
I think that's correct.
Though I would expect the seek to be of negligible performance compared to the allocation of disk space itself.

But it's always possible to do that later; if people tested that this fixes the issue let's not make further changes here.